### PR TITLE
Add support to display in JSON

### LINKED ISSUE #60: CLI JSON output support
JSON output support for all the Kadalu commands.

```
kadalu volume list --json
```

### PR DESCRIPTION
**Pull request by vatsa287**
Fixes: #60

Added JSON output with option `--json`.
- Handle errors in JSON
- Empty responses have `[]` for list & `{}` for delete/remove subcommands
- Added tests for --json

Signed-off-by: Shree Vatsa N <vatsa@kadalu.io>

### CODE REVIEW

File: mgr/src/cmds/volumes.cr [hunk 1]
```diff
@@ -3,7 +3,8 @@ require "./volume_create_parser"
 require "./gluster_volume_parser"
 
 struct VolumeArgs
-  property status = false, detail = false, name = "", volume_id = "", no_start = false, auto_create_pool = false, auto_add_nodes = false,
+  property status = false, detail = false, name = "", volume_id = "", no_start = false,
+    auto_create_pool = false, auto_add_nodes = false, json = false,
```
Comment (aravindavk) [inline mgr/src/cmds/volumes.cr]: Add `json` to Args in helpers.cr(Refer `script_mode`)

File: mgr/src/cmds/volumes.cr [hunk 2]
```diff
@@ -116,6 +117,9 @@ command "volume.list", "Volumes list of a Kadalu Storage Pool" do |parser, args|
   parser.on("--detail", "Show detailed Volumes info") do
     args.volume_args.detail = true
   end
+  parser.on("--json", "Pretty print in JSON") do
```
Comment (aravindavk) [inline mgr/src/cmds/volumes.cr]: Move this to `$SRC/mgr/src/main.cr`

File: mgr/src/cmds/volumes.cr [hunk 3]
```diff
@@ -116,6 +117,9 @@ command "volume.list", "Volumes list of a Kadalu Storage Pool" do |parser, args|
...
+    args.volume_args.json = true
```
Comment (aravindavk) [inline mgr/src/cmds/volumes.cr]: Change this to `args.json = true`

**Review by aravindavk — COMMENTED** [review]
Do not use CLI table to convert to json.

- Not all command outputs uses CLI table(For example, pool create, volume create etc)
- CLItable will have only subset of data than the actual object

Create a function to convert to json and call that from every handler

```crystal
def handle_json_output(data, args)
  if args.json
      puts data.to_json
      exit 0
  end
end
```

```crystal
pools = client.list_pools
handle_json_output(pools, args)
```

File: mgr/src/cmds/nodes.cr [hunk 4]
```diff
@@ -1,7 +1,8 @@
 require "./helpers"
+require "json"
 
 struct NodeArgs
-  property status = false, endpoint = ""
+  property status = false, endpoint = "", json = false
```
Comment (aravindavk) [inline mgr/src/cmds/nodes.cr]: No need to define this with every structs. define in common place. Check the comment above

File: mgr/src/cmds/volumes.cr (hunk 1 above)
Comment (vatsa287) [reply mgr/src/cmds/volumes.cr]: Done.

File: mgr/src/cmds/volumes.cr (hunk 3 above)
Comment (vatsa287) [reply mgr/src/cmds/volumes.cr]: Done

File: mgr/src/cmds/nodes.cr (hunk 4 above)
Comment (vatsa287) [reply mgr/src/cmds/nodes.cr]: Done

File: mgr/src/cmds/volumes.cr (hunk 2 above)
Comment (vatsa287) [reply mgr/src/cmds/volumes.cr]: Moved to main.cr

**Comment by vatsa287** [comment]
Some delete and remove subcommands do not have any responses. Should I include them as well? @aravindavk

**Comment by aravindavk** [comment]
> Some delete and remove subcommands do not have any responses. Should I include them as well? @aravindavk 

Empty json "{}" is fine for those commands.

File: mgr/src/cmds/helpers.cr [hunk 5]
```diff
@@ -105,3 +107,24 @@ def yes(label)
...
+      puts data.to_pretty_json
+      exit 0
+    else
+      puts "{}".to_json
+      exit 0
+    end
+  end
+end
+
+def handle_json_error(message, args)
```
Comment (vatsa287) [inline mgr/src/cmds/helpers.cr:122]: @aravindavk Sending err_msg in JSON. Have not tested this case.

File: mgr/src/cmds/helpers.cr [hunk 6]
```diff
@@ -105,3 +107,24 @@ def yes(label)
...
+  if args.json
+    err_msg = Hash(String, String).new
```
Comment (aravindavk) [inline mgr/src/cmds/helpers.cr]: single line shortcut `{"error": message}.to_json`

File: mgr/src/cmds/helpers.cr (hunk 5 above)
Comment (aravindavk) [reply mgr/src/cmds/helpers.cr:122]: To test socket error. Stop the Moana server and then run the CLI

File: mgr/src/cmds/volumes.cr [hunk 7]
```diff
@@ -220,7 +233,11 @@ handler "volume.delete" do |args|
   next unless (args.script_mode || yes("Are you sure you want to delete the Volume?"))
 
   api_call(args, "Failed to Delete the Volume") do |client|
-    client.pool(args.pool_name).volume(volume_name).delete
+    volume = client.pool(args.pool_name).volume(volume_name).delete
+
+    puts "volume", volume
```
Comment (aravindavk) [inline mgr/src/cmds/volumes.cr]: debug text?

File: mgr/src/main.cr [hunk 8]
```diff
@@ -31,6 +31,10 @@ module CLI
         parsed.script_mode = true if mode.strip.downcase == "script"
       end
 
+      parser_1.on("--json", "Pretty print in JSON") do |_|
```
Comment (aravindavk) [inline mgr/src/main.cr]:
```crystal
parser_1.on("--json", "Pretty print in JSON") do
    parsed.json = true
end
```

File: mgr/src/cmds/helpers.cr (hunk 6 above)
Comment (vatsa287) [reply mgr/src/cmds/helpers.cr]: updated

File: mgr/src/cmds/volumes.cr (hunk 7 above)
Comment (vatsa287) [reply mgr/src/cmds/volumes.cr]: yes missed it. removed.

File: mgr/src/main.cr (hunk 8 above)
Comment (vatsa287) [reply mgr/src/main.cr]: removed |_|.

File: mgr/src/cmds/helpers.cr [hunk 9]
```diff
@@ -105,3 +107,23 @@ def yes(label)
...
   ["yes", "y", "yy", "ok", "sure", "on"].includes?(value.strip.downcase)
 end
+
+def handle_json_output(data, args)
+  if args.json
+    if data && data != ""
+      puts data.to_pretty_json
+      exit 0
+    else
+      puts "{}".to_json
```
Comment (aravindavk) [inline mgr/src/cmds/helpers.cr]: This will print `"{}"` as string. Use `puts "{}"`

File: mgr/src/cmds/helpers.cr [hunk 10]
```diff
@@ -105,3 +107,23 @@ def yes(label)
...
+      exit 0
+    end
+  end
+end
+
+def handle_json_error(message, args)
+  if args.json
+    err = {"error": message}.to_json
+    puts err
```
Comment (aravindavk) [inline mgr/src/cmds/helpers.cr]: You can remove temp var. Directly use `puts {"error": message}.to_json`

File: mgr/src/cmds/nodes.cr [hunk 11]
```diff
@@ -88,7 +94,8 @@ handler "node.remove" do |args|
   next unless (args.script_mode || yes("Are you sure you want to remove the Node from the Pool?"))
 
   api_call(args, "Failed to Remove the Node") do |client|
-    client.pool(args.pool_name).node(node_name).delete
+    node = client.pool(args.pool_name).node(node_name).delete
+    handle_json_output(node, args)
```
Comment (aravindavk) [inline mgr/src/cmds/nodes.cr]: Remove `node` var and use `handle_json_output(nil, args)`

File: mgr/src/cmds/pools.cr [hunk 12]
```diff
@@ -48,7 +54,8 @@ handler "pool.delete" do |args|
   next unless (args.script_mode || yes("Are you sure you want to delete the Pool?"))
 
   api_call(args, "Failed to Delete the Pool") do |client|
-    client.pool(args.pool_name).delete
+    pool = client.pool(args.pool_name).delete
+    handle_json_output(pool, args)
```
Comment (aravindavk) [inline mgr/src/cmds/pools.cr]: Remove `pool` var and use `handle_json_output(nil, args)`

File: mgr/src/cmds/users.cr [hunk 13]
```diff
@@ -88,8 +94,9 @@ handler "user.logout" do |args|
   api_call(args, "Failed to Logout") do |client|
     next unless File.exists?(session_file)
 
-    client.user(client.logged_in_user_id).logout
+    user = client.user(client.logged_in_user_id).logout
     File.delete(session_file)
+    handle_json_output(user, args)
```
Comment (aravindavk) [inline mgr/src/cmds/users.cr]: Remove `user` var and use `handle_json_output(nil, args)`

File: mgr/src/cmds/users.cr [hunk 14]
```diff
@@ -103,7 +110,8 @@ handler "user.delete" do |args|
   next unless (args.script_mode || yes("Are you sure you want to delete the user(#{args.pos_args[0]})?"))
 
   api_call(args, "Failed to delete the User") do |client|
-    client.delete_user(args.pos_args[0])
+    user = client.delete_user(args.pos_args[0])
+    handle_json_output(user, args)
```
Comment (aravindavk) [inline mgr/src/cmds/users.cr]: Remove `user` var and use `handle_json_output(nil, args)`

File: mgr/src/cmds/users.cr [hunk 15]
```diff
@@ -225,19 +236,22 @@ handler "api-key.delete" do |args|
   command_error "API Key ID is required" if args.pos_args.size == 0
 
   api_call(args, "Failed to delete the API Key") do |client|
-    client.user(client.logged_in_user_id).api_key(args.pos_args[0]).delete
+    user = client.user(client.logged_in_user_id).api_key(args.pos_args[0]).delete
+    handle_json_output(user, args)
```
Comment (aravindavk) [inline mgr/src/cmds/users.cr]: Remove `user` var and use `handle_json_output(nil, args)`

File: mgr/src/cmds/volumes.cr [hunk 16]
```diff
@@ -220,7 +233,9 @@ handler "volume.delete" do |args|
   next unless (args.script_mode || yes("Are you sure you want to delete the Volume?"))
 
   api_call(args, "Failed to Delete the Volume") do |client|
-    client.pool(args.pool_name).volume(volume_name).delete
+    volume = client.pool(args.pool_name).volume(volume_name).delete
+
+    handle_json_output(volume, args)
```
Comment (aravindavk) [inline mgr/src/cmds/volumes.cr]: Remove `volume` var and use `handle_json_output(nil, args)`

File: mgr/src/cmds/helpers.cr [hunk 17]
```diff
@@ -105,3 +107,23 @@ def yes(label)
   value = STDIN.gets(chomp: true).not_nil!
   ["yes", "y", "yy", "ok", "sure", "on"].includes?(value.strip.downcase)
 end
+
+def handle_json_output(data, args)
+  if args.json
+    if data && data != ""
```
Comment (aravindavk) [inline mgr/src/cmds/helpers.cr]:
```crystal
def handle_json_output(data, args)
  return unless args.json

  if data.nil?
    puts "{}"
  else
    puts data.to_pretty_json
  end
  exit 0
end
```

**Review by aravindavk — COMMENTED** [review]
A few minor comments

File: mgr/src/cmds/helpers.cr (hunk 10 above)
Comment (vatsa287) [reply mgr/src/cmds/helpers.cr]: It was showing unexpected ":" so used a var

File: mgr/src/cmds/helpers.cr (hunk 9 above)
Comment (vatsa287) [reply mgr/src/cmds/helpers.cr]: When used `puts typeof("{}")` also prints as string, is it expected? So had used .to_json, which again prints as string.

**Comment by vatsa287** [comment]
Ack to rest of all comments regarding on passing of `nil` to empty responses.

File: mgr/src/cmds/nodes.cr (hunk 11 above)
Comment (vatsa287) [reply mgr/src/cmds/nodes.cr]: Ack to all responses on passing of `nil` for empty responses.

File: mgr/src/cmds/helpers.cr (hunk 10 above)
Comment (aravindavk) [reply mgr/src/cmds/helpers.cr]: `puts ({"error": message}.to_json)`

File: mgr/src/cmds/pools.cr [hunk 18]
```diff
@@ -13,26 +13,32 @@ handler "pool.create" do |args|
...
+
+    handle_json_output(pool, args)
+
     puts "Pool #{name} created successfully"
     puts "ID: #{pool.id}"
   end
 end
 
-command "pool.list", "Kadalu Storage Pools List" do |parser, _|
+command "pool.list", "Kadalu Storage Pools List" do |parser|
```
Comment (aravindavk) [inline mgr/src/cmds/pools.cr]: the `command` helper expects a block of two params. Is this working without the second arg?

File: mgr/src/cmds/helpers.cr (hunk 10 above)
Comment (vatsa287) [reply mgr/src/cmds/helpers.cr]: updated.

File: mgr/src/cmds/pools.cr (hunk 18 above)
Comment (aravindavk) [reply mgr/src/cmds/pools.cr]: Interestingly this is working fine even without the second parameter to the block. I added a question in the Crystal forum. I will update here once I get the clarification. https://forum.crystal-lang.org/t/how-to-validate-the-block-signature/4370

For now you can change this to

```crystal
command "pool.list", "Kadalu Storage Pools List" do |parser, _args|
```
Starting with `_` for unused args.

Comment (vatsa287) [reply mgr/src/cmds/pools.cr]: I think this was recommended by one of the linters.

**Review by aravindavk — APPROVED** [review]

**Pull request merged**